lib/core/test/unit/Cardano/Wallet/DB/Store/Transactions/StoreSpec.hs

```diff
@@ -14,23 +19,172 @@ import Cardano.Wallet.DB.Fixtures
+                    $ do
+                        Tx{..} <- select transactions
+                        txOut@(txOutPos, _) <- select $ zip [0 ..] outputs
+                        position <- get
+                        modify succ
+                        let newInput = (W.TxIn txId txOutPos, W.Coin 0)
+                        pure
+                            ( newInput
+                            ,
+                                ( mkTxIn guineaId (position, newInput)
+                                , mkTxOut (TxId txId) txOut
+                                )
+                            )
+            let guinea' = set #resolvedInputs (fst <$> txouts) guinea
+                TxHistoryF result =
+                    decorateWithTxOuts $
+                        mkTxHistory (guinea' : transactions)
+            pure (guineaId, result, txouts)
+    forAll transactionsGen $ \(idx, result, txouts) -> do
+        ins (result Map.! idx)
+            `shouldBe` [ WithTxOut txin (Just txout)
+                       | (_, (txin, txout)) <- txouts
+                       ]
+
+{- | We check that: given a set of random transactions and a guinea one for
+ which we set all collaterals to be pointing to all outputs of the others,
+ the decorated history contains a guinea transaction with the expected inputs
+-}
+prop_DecorateLinksTxCollateralsToTxOuts :: Property
+prop_DecorateLinksTxCollateralsToTxOuts = do
+    let transactionsGen = do
+            transactions :: [W.Tx] <- arbitrary
+            guinea <- arbitrary
+            let guineaId = TxId $ txId guinea
+                txouts = flip evalState 0
+                    . L.purely LT.fold L.list
+                    $ do
+                        Tx{..} <- select transactions
+                        txOut@(txOutPos, _) <- select $ zip [0 ..] outputs
+                        position <- get
+                        modify succ
+                        let newInput = (W.TxIn txId txOutPos, W.Coin 0)
+                        pure
+                            ( newInput
+                            ,
+                                ( mkTxCollateral guineaId (position, newInput)
+                                , mkTxOut (TxId txId) txOut
+                                )
+                            )
+            let guinea' =
+                    set
+                        #resolvedCollateralInputs
+                        (fst <$> txouts)
+                        guinea
+                TxHistoryF result =
+                    decorateWithTxOuts $
+                        mkTxHistory (guinea' : transactions)
+            pure (guineaId, result, txouts)
+    forAll transactionsGen $ \(idx, result, txouts) -> do
+        collateralIns (result Map.! idx)
+            `shouldBe` [ WithTxOut txin (Just txout)
+                       | (_, (txin, txout)) <- txouts
+                       ]
+
+generateLattice :: Int -> Double -> Gen (TxHistory, [W.TxIn])
```

Comment:
```suggestion
-- | ???
generateLattice :: Int -> Double -> Gen (TxHistory, [W.TxIn])
```

Brief summary of what this function does? 🤔

Reply:
I removed the generateLattice as @jonathanknowles has a shrinking solution for that in the pocket. I will postpone the idea for a while. generateLattice is just creating a valid sequence of transactions as opposed to a random bunch of them, following the utxo laws from a random genesis